torch/csrc/jit/script/compiler.cpp

```diff
@@ -29,6 +29,114 @@ using ValueTable = std::unordered_map<std::string, SugaredValuePtr>;
 using AttributeMap = std::unordered_map<std::string, Const>;
 using ListAttributeMap = std::unordered_map<std::string, std::vector<Const>>;
 
+
+using TypeAndRange = std::pair<TypePtr, const SourceRange*>;
+struct Refinements {
+
+  //using ordered map for deterministic graph output
+  std::map<std::string, TypeAndRange> mappings;
+
+  void setRefinement(const std::string& name, TypeAndRange mapping) {
+    mappings[name] = std::move(mapping);
+  }
+
+  c10::optional<TypeAndRange> getRefinement(const std::string& name) const {
+    const auto& maybe_mapping = mappings.find(name);
+    if (maybe_mapping == mappings.end()) {
+      return c10::nullopt;
+    }
+    return maybe_mapping->second;
+  }
+
+  // return the intersection of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements* intersectRefinements(const Refinements& other) {
+    Refinements ret;
+    for (auto& name_mapping: mappings) {
+      const auto& name = name_mapping.first;
+      const auto& mapping = name_mapping.second;
+      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
+        auto maybe_unified_type = unifyTypes(mapping.first, other_mapping->first);
+        if (maybe_unified_type) {
+          ret.setRefinement(name, TypeAndRange(*maybe_unified_type, mapping.second));
+        }
+      }
+    }
+    mappings = std::move(ret.mappings);
+    return this;
+  }
+
+  // return the union of the values to type mappings in a and b whose
+  // types can be unified
+  Refinements* unionRefinements(const Refinements& other) {
+    const Refinements& a = *this;
+    Refinements ret;
```

Comment:
Seems weird to construct a new Refinements just to move the result into `this`. Especially since you're copying the whole contents of `this` just to move it back.

Also, `a` is a very non-descriptive variable name, especially when you could just use `this`

Reply:
thanks will update, a was from when there was a & b